### DISCUSSION TITLE
List[int] = Form(...) -> value is not a valid integer

### DISCUSSION THREAD

**Discussion by VictorBac** [fastapi/fastapi#6861] [Questions] [answered]
Hi team,

I am trying to send a list of int alongside a Document, so I have to use Form().
The problem is that if I send a list of int, the data is passed like this:
```curl -X POST "http://localhost:8000/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d "list_int=1,2"```

I have ```-d "list_int=1,2"``` which seems to be interpreted as a float...
I get the error:
```
      "msg": "value is not a valid integer",
      "type": "type_error.integer"
```

To reproduce, just run this and call then endpoint using the OpenApi ui:
```Python
from typing import List

from fastapi import FastAPI, Form

app = FastAPI()


@app.post("/items")
async def create_item(
        list_int: List[int] = Form(...),
):
    return list_int
```

* FastAPI Version: 0.63.0

Do you have any idea how to solve this ?

**Comment by Cleptomania** [comment]
As far as I know lists/arrays aren't really a supported thing in form data in general.

I see a few potential options, one is just having separate routes and making a request for sending the file, and a separate request which can accept JSON for the other data.

Or you can somehow serialize the list at the client end into a comma separated list and just send that as a string and deserialize it on the backend.

So build a string with all of your ints like `"1,2,3,4,5"` and send that as a string value in the form data and decode it like so:

```python
from typing import List

from fastapi import FastAPI, Form

app = FastAPI()

@app.post("/items")
async def create_item(
    numbers: str = Form(...),
):
    list_ints = [int(number) for number in numbers.split(",")]
    return list_ints
```

This feels pretty hacky and not like an optimal solution though. You could instead just serialize JSON and send as a string doing something like this. Unfortunately this would not provide you with the same level of data validation that FastAPI normally provides. You would still get data validation when parsing the Pydantic object, but it probably wouldn't be as seemless as accepting a Pydantic model directly in the route.

```python
from typing import List

from fastapi import FastAPI, Form

from pydantic import BaseModel

app = FastAPI()

class Item(BaseModel):
    list_int: List[int]

@app.post("/items")
async def create_item(
    raw_data: str = Form(...),
):
    item = Item.parse_raw(raw_data)
    return item.list_int
```

With this solution you would pass something like `{"list_int": [1, 2, 3, 4]}` as the string

**Comment by ycd** [comment, marked as answer]
The method below will work just fine but as you mentioned

> To reproduce

If you have a more complex logic move that to Pydantic's validator.

**Comment by VictorBac** [comment]
Ok, I'll go with ```numbers: str = Form(...),``` for now.

Thank you guys for all the information !

**Comment by mbdaryanto** [comment]
Try to change the curl command to this
`curl -X POST "http://localhost:8000/items" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "list_int=1&list_int=2"`
Doing this, will make `list_int: List[int] = Form(...)` to receive the correct values as list of int

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @VictorBac  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
